File: java/src/main/java/io/github/athxx/ason/Parser.java
package io.github.athxx.ason;

import java.util.ArrayList;
import java.util.List;

/**
 * ASON Parser.
 */
public class Parser {
    
    private final Lexer lexer;
    private Lexer.Token current;
    
    public Parser(String input) {
        this.lexer = new Lexer(input);
        advance();
    }
    
    private void advance() {
        current = lexer.nextToken();
    }
    
    private void expect(Lexer.TokenType type) throws ParseException {
        if (current.type != type) {
            throw new ParseException("Expected " + type + ", got " + current.type +
                " at line " + current.line + ", column " + current.column);
        }
        advance();
    }
    
    public Value parse() throws ParseException {
        if (current.type == Lexer.TokenType.LBRACE) {
            return parseSchemaAndData();
        }
        return parseValue();
    }
    
    private Value parseSchemaAndData() throws ParseException {
        List<SchemaField> schema = parseSchema();
        expect(Lexer.TokenType.COLON);
        
        List<Value> results = new ArrayList<>();
        while (true) {
            Value val = parseDataWithSchema(schema);
            results.add(val);
            
            if (current.type != Lexer.TokenType.COMMA) break;
            advance();
            if (current.type != Lexer.TokenType.LPAREN) break;
        }
        
        if (results.size() == 1) {
            return results.get(0);
        }
        return Value.ofArray(results.toArray(new Value[0]));
    }
    
    private List<SchemaField> parseSchema() throws ParseException {
        expect(Lexer.TokenType.LBRACE);
        
        List<SchemaField> fields = new ArrayList<>();
        while (current.type != Lexer.TokenType.RBRACE && current.type != Lexer.TokenType.EOF) {
            fields.add(parseSchemaField());
            if (current.type == Lexer.TokenType.COMMA) {
                advance();
            }
        }
        expect(Lexer.TokenType.RBRACE);
        return fields;
    }
    
    private SchemaField parseSchemaField() throws ParseException {
        if (current.type != Lexer.TokenType.IDENT) {
            throw new ParseException("Expected identifier at line " + current.line);
        }
        
        SchemaField field = new SchemaField(current.value);
        advance();
        
        if (current.type == Lexer.TokenType.LBRACKET) {
            advance();
            field.isArray = true;
            if (current.type == Lexer.TokenType.LBRACE) {
                field.children = parseSchema();
            }
            expect(Lexer.TokenType.RBRACKET);
        } else if (current.type == Lexer.TokenType.LBRACE) {
            field.children = parseSchema();
        }
        
        return field;
    }
    
    private Value parseDataWithSchema(List<SchemaField> schema) throws ParseException {
        expect(Lexer.TokenType.LPAREN);
        
        Value obj = Value.ofObject();
        for (int i = 0; i < schema.size(); i++) {
            if (i > 0) {
                if (current.type != Lexer.TokenType.COMMA) {
                    throw new ParseException("Expected comma at line " + current.line);
                }
                advance();
            }
            
            SchemaField field = schema.get(i);
            Value val = parseFieldValue(field);
            obj.set(field.name, val);
        }
        
        expect(Lexer.TokenType.RPAREN);
        return obj;
    }
    
    private Value parseFieldValue(SchemaField field) throws ParseException {
        if (field.isArray) {
            return parseArrayValue(field);
        }
        if (field.children != null) {
            return parseDataWithSchema(field.children);
        }
        return parseValue();
    }
    
    private Value parseArrayValue(SchemaField field) throws ParseException {
        expect(Lexer.TokenType.LBRACKET);
        
        Value arr = Value.ofArray();
        boolean first = true;
        while (current.type != Lexer.TokenType.RBRACKET && current.type != Lexer.TokenType.EOF) {
            if (!first) {
                if (current.type != Lexer.TokenType.COMMA) break;
                advance();
            }
            first = false;
            
            Value val = (field.children != null) ? parseDataWithSchema(field.children) : parseValue();
            arr.push(val);
        }
        
        expect(Lexer.TokenType.RBRACKET);
        return arr;
    }
    
    private Value parseValue() throws ParseException {
        switch (current.type) {
            case NULL:
                advance();
                return Value.ofNull();
            case TRUE:
                advance();
                return Value.ofBool(true);
            case FALSE:
                advance();
                return Value.ofBool(false);
            case INTEGER:
                long intVal = Long.parseLong(current.value);
                advance();
                return Value.ofInteger(intVal);
            case FLOAT:
                double floatVal = Double.parseDouble(current.value);
                advance();
                return Value.ofFloat(floatVal);
            case STRING:
            case IDENT:
                String strVal = current.value;
                advance();
                return Value.ofString(strVal);
            case LBRACKET:
                return parseArray();
            case LPAREN:
                return parseTuple();
            default:
                throw new ParseException("Unexpected token " + current.type +
                    " at line " + current.line + ", column " + current.column);
        }
    }

    private Value parseArray() throws ParseException {
        expect(Lexer.TokenType.LBRACKET);

        Value arr = Value.ofArray();
        boolean first = true;
        while (current.type != Lexer.TokenType.RBRACKET && current.type != Lexer.TokenType.EOF) {
            if (!first) {
                if (current.type != Lexer.TokenType.COMMA) break;
                advance();
            }
            first = false;
            arr.push(parseValue());
        }

        expect(Lexer.TokenType.RBRACKET);
        return arr;
    }

    private Value parseTuple() throws ParseException {
        expect(Lexer.TokenType.LPAREN);

        Value arr = Value.ofArray();
        boolean first = true;
        while (current.type != Lexer.TokenType.RPAREN && current.type != Lexer.TokenType.EOF) {
            if (!first) {
                if (current.type != Lexer.TokenType.COMMA) break;
                advance();
            }
            first = false;
            arr.push(parseValue());
        }

        expect(Lexer.TokenType.RPAREN);
        return arr;
    }

    // Schema field helper class
    private static class SchemaField {
        String name;
        boolean isArray = false;
        List<SchemaField> children = null;

        SchemaField(String name) {
            this.name = name;
        }
    }

    // Static parse method
    public static Value parse(String input) throws ParseException {
        return new Parser(input).parse();
    }
}

File: kotlin/src/main/kotlin/io/github/athxx/ason/Parser.kt
package io.github.athxx.ason

/**
 * ASON Parser.
 */
class Parser(input: String) {
    
    private val lexer = Lexer(input)
    private var current: Lexer.Token = lexer.nextToken()
    
    private fun advance() {
        current = lexer.nextToken()
    }
    
    private fun expect(type: Lexer.TokenType) {
        if (current.type != type) {
            throw ParseException("Expected $type, got ${current.type} at line ${current.line}, column ${current.column}")
        }
        advance()
    }
    
    fun parse(): Value {
        return if (current.type == Lexer.TokenType.LBRACE) {
            parseSchemaAndData()
        } else {
            parseValue()
        }
    }
    
    private fun parseSchemaAndData(): Value {
        val schema = parseSchema()
        expect(Lexer.TokenType.COLON)
        
        val results = mutableListOf<Value>()
        while (true) {
            results.add(parseDataWithSchema(schema))
            if (current.type != Lexer.TokenType.COMMA) break
            advance()
            if (current.type != Lexer.TokenType.LPAREN) break
        }
        
        return if (results.size == 1) results[0] else Value.Array(results)
    }
    
    private data class SchemaField(
        val name: String,
        var isArray: Boolean = false,
        var children: List<SchemaField>? = null
    )
    
    private fun parseSchema(): List<SchemaField> {
        expect(Lexer.TokenType.LBRACE)
        
        val fields = mutableListOf<SchemaField>()
        while (current.type != Lexer.TokenType.RBRACE && current.type != Lexer.TokenType.EOF) {
            fields.add(parseSchemaField())
            if (current.type == Lexer.TokenType.COMMA) advance()
        }
        expect(Lexer.TokenType.RBRACE)
        return fields
    }
    
    private fun parseSchemaField(): SchemaField {
        if (current.type != Lexer.TokenType.IDENT) {
            throw ParseException("Expected identifier at line ${current.line}")
        }
        
        val field = SchemaField(current.value)
        advance()
        
        when (current.type) {
            Lexer.TokenType.LBRACKET -> {
                advance()
                field.isArray = true
                if (current.type == Lexer.TokenType.LBRACE) {
                    field.children = parseSchema()
                }
                expect(Lexer.TokenType.RBRACKET)
            }
            Lexer.TokenType.LBRACE -> {
                field.children = parseSchema()
            }
            else -> {}
        }
        
        return field
    }
    
    private fun parseDataWithSchema(schema: List<SchemaField>): Value {
        expect(Lexer.TokenType.LPAREN)
        
        val obj = Value.Object()
        schema.forEachIndexed { i, field ->
            if (i > 0) {
                if (current.type != Lexer.TokenType.COMMA) {
                    throw ParseException("Expected comma at line ${current.line}")
                }
                advance()
            }
            obj[field.name] = parseFieldValue(field)
        }
        
        expect(Lexer.TokenType.RPAREN)
        return obj
    }
    
    private fun parseFieldValue(field: SchemaField): Value {
        return when {
            field.isArray -> parseArrayValue(field)
            field.children != null -> parseDataWithSchema(field.children!!)
            else -> parseValue()
        }
    }
    
    private fun parseArrayValue(field: SchemaField): Value {
        expect(Lexer.TokenType.LBRACKET)
        
        val arr = Value.Array()
        var first = true
        while (current.type != Lexer.TokenType.RBRACKET && current.type != Lexer.TokenType.EOF) {
            if (!first) {
                if (current.type != Lexer.TokenType.COMMA) break
                advance()
            }
            first = false
            
            val item = if (field.children != null) parseDataWithSchema(field.children!!) else parseValue()
            arr.add(item)
        }
        
        expect(Lexer.TokenType.RBRACKET)
        return arr
    }
    
    private fun parseValue(): Value {
        return when (current.type) {
            Lexer.TokenType.NULL -> { advance(); Value.Null }
            Lexer.TokenType.TRUE -> { advance(); Value.Bool(true) }
            Lexer.TokenType.FALSE -> { advance(); Value.Bool(false) }
            Lexer.TokenType.INTEGER -> {
                val v = current.value.toLong()
                advance()
                Value.Integer(v)
            }
            Lexer.TokenType.FLOAT -> {
                val v = current.value.toDouble()
                advance()
                Value.Float(v)
            }
            Lexer.TokenType.STRING, Lexer.TokenType.IDENT -> {
                val v = current.value
                advance()
                Value.Str(v)
            }
            Lexer.TokenType.LBRACKET -> parseArray()
            Lexer.TokenType.LPAREN -> parseTuple()
            else -> throw ParseException("Unexpected token ${current.type} at line ${current.line}")
        }
    }

    private fun parseArray(): Value {
        expect(Lexer.TokenType.LBRACKET)

        val arr = Value.Array()
        var first = true
        while (current.type != Lexer.TokenType.RBRACKET && current.type != Lexer.TokenType.EOF) {
            if (!first) {
                if (current.type != Lexer.TokenType.COMMA) break
                advance()
            }
            first = false
            arr.add(parseValue())
        }

        expect(Lexer.TokenType.RBRACKET)
        return arr
    }

    private fun parseTuple(): Value {
        expect(Lexer.TokenType.LPAREN)

        val arr = Value.Array()
        var first = true
        while (current.type != Lexer.TokenType.RPAREN && current.type != Lexer.TokenType.EOF) {
            if (!first) {
                if (current.type != Lexer.TokenType.COMMA) break
                advance()
            }
            first = false
            arr.add(parseValue())
        }

        expect(Lexer.TokenType.RPAREN)
        return arr
    }

    companion object {
        fun parse(input: String): Value = Parser(input).parse()
    }
}

